Extension/src/LanguageServer/protocolFilter.ts

```diff
@@ -7,6 +7,29 @@
 import { Middleware } from 'vscode-languageclient';
 import { ClientCollection } from './clientCollection';
 import { Client } from './client';
+import { SourceFileConfiguration, CustomConfigurationProvider } from '../api';
+import { customConfigurationProviders } from './extension';
+
+function runThenableWithTimeout(promise: Thenable<any>, ms: number): Thenable<any> {
```

Comment:
This method should block execution.  I don't think it should return a Thenable.  It should return the configuration or something 'falsy' if it times out.

Reply:
Sorry if I'm asking a naïve question (still fairly new to JavaScript/TypeScript :) ) but how do I block on a Promise without making the current function return a Promise? It seems like if I await something, the whole stack just becomes asynchronous... Do you have any references or sample code that you can refer me to?

Reply:
You're right. I'm equally new to JavaScript and assumed that such a construct existed.  Since it doesn't look like we can block, we're probably going to need to queue up a backlog of messages and send them later when the promise is resolved.  Ugh...  client.ts does something like this with the `notifyWhenReady` and `requestWhenReady` functions.

Reply:
The Middleware only lets us intercept "textDocument/*" messages... But those would be the only messages that we would care to block any ways right?

Reply:
I'd need to think if there are any dependencies between LSP and cpptools messages.  I believe activeDocumentChanged and selectionChanged need to be sent in the correct order relative to the LSP messages to be safe.  I think we'll need to add a new method to `Client` that allows us to queue up the requests when a promise is pending.  I would expect it to be bad to infinitely chain promises off of these, so we might need to track the number of pending requests as they are serviced so we can clear the promise and go back to regular processing.  We might just need to 
make `requestWhenReady` and `notifyWhenReady` public and update them like this (not sure if this will work as written, but this is the general idea):

```
    public requestWhenReady<T>(request: () => Thenable<T>): Thenable<T> {
        if (this.pendingRequests == 0) {
            return request();
        } else if (this.isSupported && this.pendingTask) {  // this.onReadyPromise is replaced with this.pendingTask
            this.pendingRequests++;
            return this.pendingTask.then(() => {
                request();
                this.pendingRequests--;
                if (this.pendingRequests == 0) {
                    this.pendingTask = undefined;
                }
            });
        } else {
            return Promise.reject<T>("Unsupported client");
        }
    }
```

Reply:
Yeah that makes sense... Also, what about the other idea of making the server handle the delayed config notifications? Can we jam another parameter in the didOpen message to let the server know that there is a config on the way based on canProvide()? Maybe that's easier... Let me know if you want to meet up for a brainstorming session.

Reply:
didOpen is a LSP protocol message and we shouldn't tamper with that.
JavaScript is much more suited to do this delayed notifications stuff than the language server, so I would rather keep the logic in here nearby the code that has a need for it.

(i just noticed there is no logic in the code snippet above to set this.pendingTask, so it definitely won't work as written)

Reply:
Just uploaded the changes. I'm still testing it out, but thought I'd share what I have. I think all cpptools/ messages are already wrapped by notifyWhenReady or requestWhenReady, so they should be safe.